### ISSUE TITLE
call to undefined function ejabberd_oauth:verify_resowner_scope

### ISSUE THREAD

**Issue by badlop** [processone/ejabberd#2982]
In ejabberd 19.08, when trying to execute a command using OAuth and mod_http_api following [the docs example](https://docs.ejabberd.im/developer/ejabberd-api/oauth/#rest-json-example), it crashes. That worked correctly in 19.02. Crash log:

```
22:58:36.003 [info] (<0.2069.0>) Accepted connection 127.0.0.1:34986 -> 127.0.0.1:5443
22:58:36.005 [error] CRASH REPORT Process <0.2069.0>
 with 0 neighbours crashed with reason:
 call to undefined function
 ejabberd_oauth:verify_resowner_scope(
  {user,<<"user1">>,<<"localhost">>}, 
  [<<"get_roster">>,<<"sasl_auth">>], 
  {password,<<"mypass11">>})
```

That function is called from p1_oauth2:auth_user/3
https://github.com/processone/p1_oauth2/blob/master/src/oauth2.erl#L354

and this commit removed its definition:
https://github.com/processone/ejabberd/commit/a02cff0e780bb735531594c4ece81e8628f79782#diff-f7aaaa771a1b14be08cf0fae36036c6b

**Comment by zinid** [comment]
Here is the block removed:
```erl
verify_resowner_scope({user, _User, _Server}, Scope, Ctx) ->
    Cmds = ejabberd_commands:get_exposed_commands(),
    Cmds1 = ['ejabberd:user', 'ejabberd:admin', sasl_auth | Cmds],
    RegisteredScope = [atom_to_binary(C, utf8) || C <- Cmds1],
    case oauth2_priv_set:is_subset(oauth2_priv_set:new(Scope),
                                   oauth2_priv_set:new(RegisteredScope)) of
        true ->
            {ok, {Ctx, Scope}};
        false ->
            {error, badscope}
    end;
verify_resowner_scope(_, _, _) ->
    {error, badscope}.
```
The problem is that `ejabberd_commands:get_exposed_commands/0` doesn't exist anymore.
@prefiks do you know how this stuff can be rewritten?

**Comment by prefiks** [comment]
Should be fixed by commit 6d08ecc41fe236fd9effb08ad784f95caadc9c5b

**Comment by jemliF** [comment]
I am using 19.08 version and still having the same issue.

> [error] CRASH REPORT Process <0.912.0> with 0 neighbours crashed with reason: call to undefined function ejabberd_oauth:verify_resowner_scope({user,<<"test">>,<<"157.245.128.100">>}, [<<"get_roster">>,<<"sasl_auth">>], {password,<<"testtt">>})
> 12:15:15.142 [error] Supervisor ejabberd_http_sup had child undefined started with {ejabberd_http,start_link,undefined} at <0.912.0> exit with reason call to undefined function ejabberd_oauth:verify_resowner_scope({user,<<"test">>,<<"157.245.128.100">>}, [<<"get_roster">>,<<"sasl_auth">>], {password,<<"testt">>}) in context child_terminated

**Comment by zinid** [comment]
@jemliF because it was introduced in 19.08. The fix will be in 19.09 coming next week.

**Comment by m-faraz** [comment]
Is this fixed now? I see no mention of this in the changelogs/commits for both 19.09 and 19.09.1!

**Comment by prefiks** [comment]
Yes, this fix is in 19.09
